### PR TITLE
primitives/chain: dev chain support cancun upgrade

### PR DESCRIPTION
**Pull request by jsvisa**
Run `node --dev` support Cancun upgrade

### CODE REVIEW

File: crates/primitives/src/chain/spec.rs [hunk 1]
```diff
@@ -235,6 +235,7 @@ pub static DEV: Lazy<Arc<ChainSpec>> = Lazy::new(|| {
                 ForkCondition::TTD { fork_block: Some(0), total_difficulty: U256::from(0) },
             ),
             (Hardfork::Shanghai, ForkCondition::Timestamp(0)),
+            (Hardfork::Cancun, ForkCondition::Timestamp(0)),
```
Comment (mattsse) [inline crates/primitives/src/chain/spec.rs:238]: makes sense, please also add the cancun() to fork_timestamps above to L219

**Review by mattsse — CHANGES_REQUESTED** [review]
makes sense

File: crates/primitives/src/chain/spec.rs (hunk 1 above)
Comment (jsvisa) [reply crates/primitives/src/chain/spec.rs:238]: thanks for that, seems we also need to adjust something in the auto-seal crate, currently the `parent_beacon_block_root` is null when inserting a blob tx:


```
2024-03-07T22:50:23.063877Z  WARN Failed to insert downloaded block err=Failed to insert block (hash=0xb0a303a752869d4d128a9901784a31fd1242f31cbdd570df6891f07834192ac1, number=1, parent_hash=0x2f980576711e3617a5e4d83dd539548ec0f7792007d505a3d2e9674833af2d7c): EIP-4788 parent beacon block root missing for active Cancun block
2024-03-07T22:50:23.063894Z  WARN Marking block as invalid invalid_number=1 invalid_hash=0xb0a303a752869d4d128a9901784a31fd1242f31cbdd570df6891f07834192ac1 err=EIP-4788 parent beacon block root missing for active Cancun block
2024-03-07T22:50:23.063901Z  WARN Bad block with hash hash=0xb0a303a752869d4d128a9901784a31fd1242f31cbdd570df6891f07834192ac1 header=Header { parent_hash: 0x2f980576711e3617a5e4d83dd539548ec0f7792007d505a3d2e9674833af2d7c, ommers_hash: 0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347, beneficiary: 0x0000000000000000000000000000000000000000, state_root: 0x271ddc29ac3d4b22dcb626913d1a5c04c1ae3a11cd27d4b6778facbd8d58fb22, transactions_root: 0x4d4769da8d6bfb40af858b494e7e9b4e3b109b582dcb643abb2ba6c45693db7c, receipts_root: 0xeaa8c40899a61ae59615cf9985f5e2194f8fd2b57d273be63bde6733e89b12ab, withdrawals_root: None, logs_bloom: 0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000, difficulty: 0x0000000000000000000000000000000000000000000000000000000000000002_U256, number: 1, gas_limit: 30000000, gas_used: 21000, timestamp: 1709851823, mix_hash: 0x0000000000000000000000000000000000000000000000000000000000000000, nonce: 0, base_fee_per_gas: Some(875000000), blob_gas_used: None, excess_blob_gas: None, parent_beacon_block_root: None, extra_data: 0x }
2024-03-07T22:50:23.063962Z ERROR Forkchoice update returned invalid response fcu_response=OnForkChoiceUpdated { forkchoice_status: Invalid, fut: Left(Ready(Some(Ok(ForkchoiceUpdated { payload_status: PayloadStatus { status: Invalid { validation_error: "links to previously rejected block" }, latest_valid_hash: Some(0x2f980576711e3617a5e4d83dd539548ec0f7792007d505a3d2e9674833af2d7c) }, payload_id: None })))) }
2024-03-07T22:50:23.063978Z  INFO Forkchoice updated head_block_hash=0xb0a303a752869d4d128a9901784a31fd1242f31cbdd570df6891f07834192ac1 safe_block_hash=0xb0a303a752869d4d128a9901784a31fd1242f31cbdd570df6891f07834192ac1 finalized_block_hash=0xb0a303a752869d4d128a9901784a31fd1242f31cbdd570df6891f07834192ac1 status=Invalid

```

**Review by mattsse — DISMISSED** [review]

File: crates/primitives/src/chain/spec.rs (hunk 1 above)
Comment (mattsse) [reply crates/primitives/src/chain/spec.rs:238]: right, I think we're missing either pre or post block cancun call, need to check and parent beacon block hash

will check

Comment (jsvisa) [reply crates/primitives/src/chain/spec.rs:238]: fix it.

**Comment by jsvisa** [comment]
Could you please take another look @mattsse . The blob blocks now can be successfully build, but the block’s blob gas used was not calculated(maybe i’ll file in another pr)

**Comment by jsvisa** [comment]
@mattsse sorry to bother you, but is there anything more I need to adjust?

**Review by mattsse — APPROVED** [review]
thanks!

**Pull request merged**